_posts/2022-01-01-the-kubernetes-discovery-cache-blessing-and-curse.md

```diff
@@ -0,0 +1,101 @@
+
+It all started for me with a tweet by Bryan Liles ([VP, Principal Engineer at VMWare](https://www.linkedin.com/in/bryanliles/)) almost a year ago, of which Tim Hockin ([Principal Software Engineer at Google & Founder of Kubernetes](https://www.linkedin.com/in/tim-hockin-6501072/)) was [affirmative](https://twitter.com/thockin/status/1346313102430670851):
+
+<blockquote class="twitter-tweet tw-align-center"><p lang="en" dir="ltr">I often think about Kubernetes and how it rubs people the wrong way. There are lots of opinions about k8s as a container orchestrator, but the real win (and why I think it is great) is the core concept: Platform for managing declarative APIs. Infra APIs have never been better.</p>&mdash; Bryan Liles (@bryanl) <a href="https://twitter.com/bryanl/status/1346125863419568129">January 4, 2021</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
+
+I thought this was so true. Kubernetes embodies a lot of principles of the perfect API. Every resource has a group, a version and a kind, it offers extensible metadata, a spec for input data (desired state) and a status of the resource (actual state). What else do you wish for? Also the core Kubernetes controllers like the deployment controller, pod controller, autoscaler controller and persistent volume claim controller are perfect examples of asynchronous controller patterns that take a desired state as input and achieve that state through reconciliation with eventual consistency. The fact that this functionality is exposed to Kubernetes users via custom resource definitions (CRDs) makes the entire platform incredibly extensible.
+
+Controllers like [crossplane.io](https://crossplane.io), [GCP Config Connector](https://github.com/GoogleCloudPlatform/k8s-config-connector) or [Azure Service Operator](https://github.com/Azure/azure-service-operator) have adopted the pattern to a large degree and install 100s, if not 1,000s of CRDs on clusters. However that doesn't come without its drawbacks...
+
+<!--more-->
+
+These draw backs aren't due to high load on the Kubernetes API server. In fact, that actually has a pretty robust and advanced rate limiting mechanism through the [priority and fairness](https://kubernetes.io/docs/concepts/cluster-administration/flow-control) design, that most likely will ensure that the API server doesn't crash even if a lot of requests are made.
+
+However, installing many CRDs on a cluster can impact the [OpenAPI spec publishing](https://github.com/crossplane/crossplane/issues/2649), as well as the [discovery cache creation](https://github.com/kubernetes/kubectl/issues/1126). OpenAPI spec creation for many CRDs has [recently been fixed](https://github.com/kubernetes/kube-openapi/pull/251) through implementing lazy marshalling. While an interesting concept, this could be the topic of another blog post as in this one we are focusing on the latter one: discovery cache creation.
```

Comment:
```suggestion
However, installing many CRDs on a cluster can impact the [OpenAPI spec publishing](https://github.com/crossplane/crossplane/issues/2649), as well as the [discovery cache creation](https://github.com/kubernetes/kubectl/issues/1126). OpenAPI spec creation for many CRDs has [recently been fixed](https://github.com/kubernetes/kube-openapi/pull/251) through implementing lazy marshalling. While an interesting concept, this could be the topic of another blog post as in this one we are focusing on the latter: discovery cache creation.
```